core/trino-main/src/main/java/io/trino/execution/scheduler/FixedCountNodeAllocator.java

```diff
@@ -0,0 +1,206 @@
+                .collect(toImmutableList());
+
+        if (nodesMatchingRequirements.isEmpty()) {
+            throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+        }
+
+        Optional<InternalNode> selectedNode = nodesMatchingRequirements.stream()
+                .filter(node -> allocationCountMap.getOrDefault(node, 0) < maximumAllocationsPerNode)
+                .min(comparing(node -> allocationCountMap.getOrDefault(node, 0)));
+
+        if (selectedNode.isEmpty()) {
+            return Optional.empty();
+        }
+
+        int allocationCount = allocationCountMap.compute(selectedNode.get(), (key, value) -> value == null ? 1 : value + 1);
+        checkState(allocationCount <= maximumAllocationsPerNode, "allocation count for node %s is expected to be less than %s: %s", selectedNode.get(), maximumAllocationsPerNode, allocationCount);
+        return selectedNode;
+    }
+
+    private synchronized void releaseNodeInternal(InternalNode node)
+    {
+        int allocationCount = allocationCountMap.compute(node, (key, value) -> value == null ? 0 : value - 1);
+        checkState(allocationCount >= 0, "allocation count for node %s is expected to be greater than or equal to zero: %s", node, allocationCount);
+    }
+
+    private void processPendingAcquires()
+    {
+        verify(!Thread.holdsLock(this));
+
+        Map<PendingAcquire, InternalNode> assignedNodes = new IdentityHashMap<>();
+        Map<PendingAcquire, RuntimeException> failures = new IdentityHashMap<>();
+        synchronized (this) {
+            Iterator<PendingAcquire> iterator = pendingAcquires.iterator();
+            while (iterator.hasNext()) {
+                PendingAcquire pendingAcquire = iterator.next();
+                if (pendingAcquire.getFuture().isCancelled()) {
+                    iterator.remove();
+                    continue;
+                }
+                try {
+                    Optional<InternalNode> node = tryAcquireNode(pendingAcquire.getNodeRequirements());
+                    if (node.isPresent()) {
+                        iterator.remove();
+                        assignedNodes.put(pendingAcquire, node.get());
+                    }
+                }
+                catch (RuntimeException e) {
+                    iterator.remove();
+                    failures.put(pendingAcquire, e);
+                }
+            }
+        }
+
+        assignedNodes.forEach((pendingAcquire, node) -> {
+            SettableFuture<InternalNode> future = pendingAcquire.getFuture();
+            future.set(node);
+            if (future.isCancelled()) {
+                releaseNodeInternal(node);
+            }
+        });
+
+        failures.forEach((pendingAcquire, failure) -> {
+            SettableFuture<InternalNode> future = pendingAcquire.getFuture();
+            future.setException(failure);
+        });
+    }
+
+    @Override
+    public synchronized void close()
+    {
+    }
```

Comment:
feels like you should release all aquired nodes here. Or check that all were release before close.

Reply:
This implementation only maintains state per query. Once we have an implementation that maintains a global state (between different queries) the close can be used to do the cleanups.